### PR TITLE
Show inherited item descriptions

### LINKED ISSUE #88: For an overridden method with no documentation, automatically show the description from ancestor, clearly marking that "this is a virtual method defined in ancestor, showing description from ancestor"

### PR DESCRIPTION
**Pull request by fourls**
Rewrite description handling to show inherited descriptions of the item from ancestor classes/interfaces in HTML output.

The inherited descriptions that will be shown are from:

- all interfaces implemented by this class, whether directly or via inheritance
- the first ancestor class with a description for this item

If there are no descriptions to inherit, show "This item has no description." As it is very easy to hide this using CSS, having this output provides more flexibility than not outputting it at all.

Resolves #88. 

### Examples with default CSS

No description

image

Single inherited description

image

Multiple inherited descriptions

image

### CODE REVIEW

**Comment by michaliskambi** [comment]
Great, thank you for doing this! This was an important feature we had missing in PasDoc, I also missed it in my Castle Game Engine ( https://github.com/castle-engine/castle-engine/issues/177 ) but never had time to do it in PasDoc.

I'll review it this weekend. Functionality that you show (gathering from ancestors and interfaces) looks great.

File: source/component/PasDoc_Items.pas [hunk 1]
```diff
@@ -1705,6 +1731,88 @@ procedure TPasItem.StoreParamTag(
...
+  else
+    Result := nil;
+end;
+
+function TPasItem.GetInheritedItemDescriptions: TStringPairVector;
+
+  function IsOverrideItem(Item: TPasItem): Boolean;
+  begin
+    Result := (
+     ((Item is TPasRoutine) and (SD_OVERRIDE in TPasRoutine(Self).Directives)) or
```
Comment (michaliskambi) [inline source/component/PasDoc_Items.pas:1747]: I think above you meant `TPasRoutine(Item)` not `TPasRoutine(Self)`. 

Ah, but I see it doesn't matter -- you always call `IsOverrideItem(Self)`.

Another idea then: make IsOverrideItem a protected virtual method in TPasItem. From what I understand, you can make it functionally equivalent to current code, but avoid "is" checks and typecasts:

- Let it just return false in TPasItem.
- In TPasRoutine let it return `SD_OVERRIDE in Directives`
- In TPasProperty let it return `PropType = ''`

File: source/component/PasDoc_Items.pas [hunk 2]
```diff
@@ -1705,6 +1731,88 @@ procedure TPasItem.StoreParamTag(
...
+  begin
+    CurrentClassAncestor := MyObject;
+    while Assigned(CurrentClassAncestor) do
+    begin
+      { We want the main ancestor's description above all else, unless the method is not in
+        the inheritance tree of the method anymore. We want to keep traversing past that point
+        because we want to pick up descriptions from any interfaces. }
+      if not OverrideChainEnded then
+      begin
+        ThisItemInAncestor := CurrentClassAncestor.FindItem(Self.Name) as TPasItem;
```
Comment (michaliskambi) [inline source/component/PasDoc_Items.pas:1780]: Write just `Name`, not `Self.Name`. We generally avoid writing explicit `Self` when it is not necessary.

**Review by michaliskambi — CHANGES_REQUESTED** [review]
Thank you! This looks great. I found 2 small things to improve (see comments above).

As an addition, bigger request: please add a testcase for this feature:

- add a unit `tests/testcases/ok_inherited_descriptions.pas` that exercises various points of this logic. 
- add to `tests/scripts/mk_tests.sh` a line calling it: 
```
  mk_test ok_inherited_descriptions ok_inherited_descriptions.pas
```
- run the tests by entering `tests` and executing `make`. And see if results for ok_inherited_descriptions.pas look good. This requires some command-line setup (see tests/README.md) -- you can leave this to me if you want. If you add a testcase, I'll test it and commit valid results.

**Comment by fourls** [comment]
Thanks for reviewing! I'll push an updated version sometime soon. 

In regards to tests, I'm currently using Delphi on a Windows machine so I'll have to set some things up - I'll let you know if I have any issues.

**Comment by fourls** [comment]
I've just pushed the two small improvements - I haven't got to the tests yet.

**Comment by fourls** [comment]
I've written a test and regenerated the test cases. There are a lot of changes! This is mostly due to now explicitly stating when an item has no description. Could you double check that the testing output is complete? (I don't have PHP on my system and the test script fails when it gets to PHP validation.)

I also realised that the CIO descriptions were only being directly inherited instead of traversing the tree, so I've fixed that as well.

**Comment by michaliskambi** [comment]
Sorry for the delay with reviewing -- I have a bit of busy time (developing 4 features in parallel in https://castle-engine.io/ :) ).

I have this in TODO and will definitely look at it this week.

**Comment by fourls** [comment]
No problem, thanks Michalis!

**Comment by fourls** [comment]
Hi, just wondering if you've had time to take a look at this yet?

**Comment by michaliskambi** [comment]
Sorry, I failed behind on my tasks, trying to do too many things at once :) This Wednesday will be my day for handling all PasDoc PRs, promised!

**Review by michaliskambi — APPROVED** [review]

**Comment by michaliskambi** [comment]
Thank you very much, and sorry for the delay in reviewing. Please continue contributing to PasDoc and don't be discouraged by my delays in reviewing :)

Overall: Excellent work, thank you!

1. Good job unifying CIOs and methods - I see the `WriteNoDescription` is now uniform approach, using in turn virtual `GetInheritedItemDescriptions`.

    And we even have better error messages now, e.g. I see using `@inherited` on a global routine now returns

    ```
    This item is not a member of a class/interface/etc. (in description of "warning_inherited_test.Test")
    ```

    instead of previous

    ```
    Can't resolve @inherited: Member "Test" not found in ancestors (in description of "warning_inherited_test.Test")
    ```

    (which wasn't sensible as Test was a global routine, it has no concept of ancestors.)

2. The huge difference on `testcases_output` is indeed unfortunately unavoidable for such changes. I scanned it quickly to see that all changes seem to be in line with what we wanted, no regressions.

3. ... Although the huge difference on all PHP files was avoidable, it was our mistake -- the PHP files were not marked properly as text files in `.gitattributes`, and in effect your test on Windows generated them all with Windows line endings (`#13#10`), in repo we had Unix line endings (`#10`), and GIT thought that it means that every line of every PHP file is now different.

    This is fixed now, i.e. I fixed `.gitattributes` and recommitted PHP files (making sure that the only difference was indeed in line endings using `git diff -w` and `diff -wur testcases_output/php/ before_testcases_output/php/`). So this problem will not return.

4. I noticed that the cache test failed. This was visible when doing `make` in `tests/` subdirectory or executing shell script `./check_cache.sh html` in `tests/scripts/` subdirectory. I attach the output.

    [check_cache_output.txt](https://github.com/pasdoc/pasdoc/files/10026916/check_cache_output.txt)

    No worries, I know that using these scripts is a bit painful on Windows :) I should set up GitHub Actions to automatically test all PRs, then such failure will be easily visible, and without the need to set up anything for each developer. I really enjoy GitHub Actions (see https://castle-engine.io/github_actions), I'm adding to my TODO to set them up for PasDoc.

    I fixed this, so this is just a note for future, so you note what it means:

    The `./check_cache.sh xxx` checks that our cache (see https://pasdoc.github.io/CacheOption ) works OK. In the 1st pass, it generates docs and writes the cache. In the 2nd pass, it reads the cache and generates the docs. The resulting docs should be exactly equal (IOW, using the cache is only an optimization that doesn't affect how the result looks like). The error, in this case on `ok_inherited_descriptions.pas` (your new testcase) says that we probably don't serialize / deserialize everything we should to have the same output on 2nd run.

    So I cross-checked what you access in `GetInheritedItemDescriptions` for routines, with what is serialized, and found we didn't serialize/deserialize `TPasRoutine.Directives`, and your `TPasRoutine.IsOverride` depends on it to make decisions. As it happened, we didn't detect previously the fact that it's not serialized properly, no other testcase needed that.

    After https://github.com/pasdoc/pasdoc/commit/b15ce4cdc5f422c2ec3eb0d34af352b1c0634135 it's all good, tests pass 100%.

6. After some thinking and experimenting, I decided to make `.nodescription` items hidden by default.

    Reason: the sentence "This item has no description." seemed a bit unnecessary for me when we don't follow it by any description from the ancestor. It meant that processing an undocumented unit like this:

    ```pascal
    unit b;
    interface
    type
      TMyClass = class
        procedure Foo;
      end;
    procedure Bar;
    implementation
    end.
    ```

    shows `<p class="nodescription">This item has no description.</p>` for the unit, class, Foo and Bar.

    This information

    - can be seen as valid, moreover useful informaiton -> it's encouraging developer to actually add the documentation there.

    - But then, it can also be seen as "too much". I'm sometimes OK with leaving some identifiers undocumented in the (admittedly rare) cases when identifier name really says everything (see CGE coding conventions, https://castle-engine.io/coding_conventions#_document_at_least_every_protected_public_published_identifier_use_a_language_that_looks_good_in_english_and_uses_pasdoc_tags ) so this sentence would be undesired.

    Opinions (of everyone) welcome, we can change it back (and particular projects can of course hide `.nodescription` by their custom CSS if needed). Right now, if you want to see it, then enable it in CSS by `.nodescription { display: block; }` in CSS. Or look at HTML sources :)

**Comment by fourls** [comment]
Thanks Michalis! 

> I noticed that the cache test failed. This was visible when doing make in tests/ subdirectory or executing shell script ./check_cache.sh html in tests/scripts/ subdirectory.

Ah, I might not have noticed this because `validate_php.sh` failed due to not having a PHP install on my system. Thanks for fixing it up. I hadn't noticed this in any of my manual testing because I don't use the cache feature. 

> I should set up GitHub Actions to automatically test all PRs, then such failure will be easily visible, and without the need to set up anything for each developer. I really enjoy GitHub Actions (see https://castle-engine.io/github_actions), I'm adding to my TODO to set them up for PasDoc.

Sounds like a great idea! I love the convenience of GitHub Actions as well, I've found it to be useful on pull requests in the past since it's like immediately getting a basic review of the PR.

> After some thinking and experimenting, I decided to make `.nodescription` items hidden by default.

Makes perfect sense to me. I'm not opinionated on the default, I just wanted the choice to be available :)

Thanks for merging - no worries about the wait, looks like you're doing good work on Castle Engine. I expect to be contributing more to PasDoc in the near future :)

**Comment by Fr0sT-Brutal** [comment]
> * CGE coding conventions

> We detect Delphi by just lack of FPC

There's `DCC` define appeared in XE2. [SO](https://stackoverflow.com/questions/21175071/delphi-equivalent-to-free-pascals-fpc-define). For earlier versions I've added
```
// Define DCC if not defined yet (Delphi up to XE2)
{$IFNDEF DCC}
  {$IF DECLARED(CompilerVersion)}
    {$DEFINE DCC}
  {$IFEND}
{$ENDIF}
```

In theory, there's some other compilers than Delphi & FPC (PascalABC, Oxygen, etc). It's better to define Delphi-specific pieces with explicit Delphi define

**Pull request merged**